splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/MultiProbeDerbyScanInformation.java

```diff
@@ -293,39 +301,64 @@ public void readExternal(ObjectInput in) throws IOException, ClassNotFoundExcept
 					continue;
 
 				int maxSize = dtd.getMaximumWidth();
+				DataValueDescriptor targetDesc = templateRow.getColumn(keyTablePositionMap[keyDecodingMap[position]] + 1);
+				boolean targetIsSQLVarcharDB2Compatible = targetDesc instanceof SQLVarcharDB2Compatible;
 				for (int index = 0; index < probingVals.length; index++) {
 					 DataValueDescriptor dvd = probingVals[index];
 					 if (dvd instanceof ListDataType) {
 						 ListDataType listData = (ListDataType) dvd;
 						 DataValueDescriptor dvd1 = listData.getDVD(position);
-						 if (dvd1 instanceof SQLChar) {
+						 boolean isSQLVarcharDB2Compatible = dvd1 instanceof SQLVarcharDB2Compatible;
+						 if (isSQLVarcharDB2Compatible) {
+							DataValueDescriptor newDVD =
+								QualifierUtils.convertChar((SQLChar) dvd1, dvd1.getLength(),
+								                           inlistDataTypes[position].getMaximumWidth());
+							listData.setDVD(position, newDVD);
+						 }
+						 else if (dvd1 instanceof SQLChar) {
 						 	// we may prune some probe value based on the string length
-						 	if (isFixedCharType && dvd1.getLength() != maxSize) {
-								toRemove[index] = true;
-								continue;
-							} else if (isVarCharType && dvd1.getLength() > maxSize) {
+						 	if (isVarCharType && dvd1.getLength() > maxSize &&
+							    !targetIsSQLVarcharDB2Compatible) {
 								toRemove[index] = true;
 								continue;
 							}
 							 // if column is of varchar type, we need to change the probe values from SQLChar to SQLVarchar,
-							 // so that duplicate removal won't ignore the trailing spaces
-							 if (isVarCharType && !(dvd1 instanceof SQLVarchar))
-								 listData.setDVD(position, new SQLVarchar(dvd1.getString()));
+							 // so that duplicate removal won't ignore the trailing spaces,
+							 // unless we're in DB2 varchar compatibility mode.
+							 if (isVarCharType) {
+							 	if (targetIsSQLVarcharDB2Compatible) {
+							 		if (dvd1.getClass().equals(SQLVarchar.class))
+							 			listData.setDVD(position,
+										                new SQLVarcharDB2Compatible(dvd1.getString()));
+								}
+							 	else if (!(dvd1 instanceof SQLVarchar))
+								     listData.setDVD(position, new SQLVarchar(dvd1.getString()));
+							 }
```

Comment:
same code as 347-361. i think we should combine those in a function

Reply:
+1 for this

Reply:
OK, I've refactored this into a function.